articles/qos_deadline_liveliness_lifespan.md

```diff
@@ -0,0 +1,241 @@
+author: '[Nick Burek](https://github.com/nburek)'
+published: true
+categories: Middleware
+date: February 13th 2019
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+Glossary:
+
+- DDS - Data Distribution Service
+- RTPS - Real-Time Publish Subscribe
+- QoS - Quality of Service
+- Client - Refers to an application that connects to a ROS Service to send requests and receive responses.
+- Owner - Refers to the application that is running a ROS Service that receives requests and sends responses.
+
+## Existing ROS QoS Settings
+
+While DDS provides many settings to enable fine-grained control over the Quality of Service (QoS) for entities, ROS only provides native support for a handful of them. ROS users are able to specify the History, Depth, Reliability, and Durability via a QoS configuration struct when they create Publishers, Subscribers, etc.
+
+This leaves a lot of QoS settings that can only be set if DDS vendor can load additional default settings via a configuration file.
+If a user wants to hook their code into these additional QoS settings then they would need to get a reference to the rmw implementation and program against the vendor specific API.
+Without the abstraction layer provided by ROS their code becomes less portable.
+
+## New QoS Settings
+
+As users start to build more robust applications, they are going to need more control over when data is being delivered and information about when the system is failing.
+To address these needs it was proposed that we start by adding support for the following new DDS QoS settings.
+
+### Deadline
+
+The deadline policy establishes a contract for the amount of time allowed between messages.
+For Topic Subscribers it establishes the maximum amount of time allowed to pass between receiving messages.
+For Topic Publishers it establishes the maximum amount of time allowed to pass between sending messages.
+For Service Owners it establishes the maximum amount of time allowed to pass between receiving a request and when a response for that request is sent.
+For Service Clients it establishes the maximum amount of time allowed to pass between sending a request and when a response for that request is received.
+
+Topics and Services will support the following levels of deadlines.
+- DEADLINE_DEFAULT - Use the ROS specified default for deadlines (which is DEADLINE_NONE).
+- DEADLINE_NONE - No deadline will be offered or requested and events will not be generated for missed deadlines.
+- DEADLINE_RMW - The rmw layer of ROS will track the deadline. For a Publisher or Subscriber this means that a deadline will be considered missed if the rmw layer has not received a message within the specified time. For a Service this means the time a request is started is marked when the request reaches the rmw layer and the time at which it is finished is when the response message reaches the rmw layer.
```

Comment:
I'm happy to change the name if someone can provide me with a better suggestion. I was trying to find a name that captured the fact that times are recorded in the rmw layer and leave it open to the possibility of a setting in the future that provides a more full stack deadline without changing the behavior of existing options. 

I understand what you're saying about exposing implementation details, but at the same time I don't want to abstract away information that a user trying to fine tune their system will find useful. The design of ROS2 does expose the architecture of having the library built on top of the rmw layer by allowing users to specify what rmw implementation they want to run. So I don't feel like this name exposes any more implementation detail of the existence of the rmw layer than that does.

Reply:
What about "DEADLINE_MW" for "middleware"? It's less specific to `rmw` while still indicating that what you use to communicate gets to decide.

Reply:
> For a Publisher or Subscriber this means that a deadline will be considered missed if the rmw layer has not received a message within the specified time.

I can't find this quickly now but does ROS 2 Subscription own and manage DataReaders (and likewise ROS 2 Publisher for DataWriters)? If yes, is there a 1-1 relationship ROS 2 Subscription-DataReader? If no and if ROS 2 Subscription owns and manages multiple DataReaders then we shall find out which DataReader missed the deadline, no?

Reply:
I agree with @wjwwood that the developer should not have to deal with the implementation details of the underlying middleware. I think the QOS policy should just be a duration for the deadline and away (either by a bool or an infinite duration) to indicate that no deadline checking should occur.

Reply:
@dejanpan: Yes, there is a 1-1 relationship ROS 2 Subscription-DataReader.

Reply:
> Service this means the time a request is started is marked when the request reaches the rmw layer and the time at which it is finished is when the response message reaches the rmw layer.

Will the deadline for Services then also indirectly account for the user specified `timeout`  in the sync service Client? I feel that the `deadline` and `timeout` in this case will have an overlap and it could become confusing for the user how to combine them.

With this we also deviate from topics where deadline is used for one-way communication to services where we use deadline for a round trip. And I do not really see a rationale for it.